### PR TITLE
wait-free relaxedPoll for mpmc xadd q

### PR DESCRIPTION
**Pull request by franz1981**
it provides a full-fat wait-free `relaxedPoll` for the mpmc xadd q.
Soon I will publish some results to check how/if it improves if compared with `poll`.
I'm not sure I will provide `relaxedPeek` unless strictly necessary.

My concern is that given the many indirections of this queue, a relaxedPoll won't give any perf benefit if not to be sure that the operation completes in a finite number of steps (as a wait-free one is expecting to do).

### CODE REVIEW

File: jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java [hunk 1]
```diff
@@ -658,7 +658,97 @@ public boolean relaxedOffer(E e)
...
+            if (pooled)
+            {
+                if (next.lvSequence(consumerOffset) != chunkIndex)
+                {
+                    return null;
+                }
+            }
+            else
+            {
+                e = next.lvElement(consumerOffset);
```
Comment (franz1981) [inline jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java]: @nitsanw We don't know here if `next::index == chunkIndex`, but we don't care as long as `e !=null` and we succeed to increment `consumerIndex`

**Comment by franz1981** [comment]
@nitsanw By looking at the code I see that a slow consumer could `lvNext` while another would `spNext(null)` on the same `consumerBuffer`: it shouldn't be a big deal given that `casConsumerIndex` will protect the former to perform interesting logic on `consumerBuffer`, but for atomicity/correctness would be better to turn the `spNext` into a `soNext`, wdyt?

It won't save navigating `consumerBuffer::next` first element if `next != null', but will make sure the `next` read will perform atomically.

**Comment by franz1981** [comment]
@nitsanw I've added a new commit to discuss a different way to check if a chunk is correct. 
TLDR 
In the original version on `rotateConsumerBuffer`:
```java
consumerBuffer.soIndex(AtomicChunk.NIL_CHUNK_INDEX);
//...
consumerBuffer.spNext(null);
//..
```
While on `poll` of `firstElementOfNewChunk`:
```java
next = consumerBuffer.lvNext();
final long expectedChunkIndex = chunkIndex - 1;
if (expectedChunkIndex != consumerBuffer.lvIndex())
{
    continue;
}
```
In my original intent the order of checks was to be sure that the `next` value was still valid to be used (ie `consumerBuffer` wasn't marked as a *NIL* chunk) failing fast without fetching the element on next and check for q emptyness.
In later versions of `rotateConsumerBuffer` I've removed `consumerBuffer.soIndex(AtomicChunk.NIL_CHUNK_INDEX)`, hence I don't see a reason now to maintain the same order of checks on `poll`, given that:

1. `casConsumerIndex` already protect racing rotating consumers which `consumerBuffer::next` is `null` because of the winning rotating consumer
2. it would optimize the common case for a rare event

In the new version on `rotateConsumerBuffer`:
```java
consumerBuffer.soNext(null);
```
While on `poll` of `firstElementOfNewChunk`:
```java
final long expectedChunkIndex = chunkIndex - 1;
if (expectedChunkIndex != consumerBuffer.lvIndex())
{
    continue;
}
next = consumerBuffer.lvNext();
```
I'm now running some benchmark to see if It won't introduce any perf regression, but don't merge it please: I need your opinion on the logic.

**Comment by franz1981** [comment]
I cannot see evident regressions on `QueueBurstCost`, but in the 3P-3C case, I'm running it again to be sure of it.

File: jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java [hunk 2]
```diff
@@ -541,15 +548,18 @@ public E poll()
                         continue;
                     }
                 } else {
-                    e = consumerBuffer.lvElement(consumerOffset);
                     final long index = consumerBuffer.lvIndex();
-                    if (index != chunkIndex || e == null)
+                    if (index != chunkIndex || (e = consumerBuffer.lvElement(consumerOffset)) == null)
```
Comment (nitsanw) [inline jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java]: Prev style is clearer IMO

File: jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java [hunk 3]
```diff
@@ -532,6 +535,10 @@ public E poll()
                             //stale view of the world
                             continue;
                         }
+                        //it cover both cases:
+                        //- right chunk, awaiting element to be set
+                        //- old chunk, awaiting rotation
+                        //It allows to fail fast if the q is empty after the first element on the new chunk.
```
Comment (nitsanw) [inline jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java]: Inconsistent capitalization, make up your mind :)

File: jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java [hunk 4]
```diff
@@ -658,7 +668,96 @@ public boolean relaxedOffer(E e)
...
-        return poll();
+        final int chunkMask = this.chunkMask;
+        final int chunkShift = this.chunkShift;
+        final int chunkSize = chunkMask + 1;
+        final long consumerIndex = this.lvConsumerIndex();
+        final AtomicChunk<E> consumerBuffer = this.lvConsumerBuffer();
+        final int consumerOffset = (int) (consumerIndex & chunkMask);
+        final long chunkIndex = consumerIndex >> chunkShift;
+        final boolean firstElementOfNewChunk = consumerOffset == 0 && consumerIndex >= chunkSize;
+        if (firstElementOfNewChunk)
```
Comment (nitsanw) [inline jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java]: Surely this is rare and can be extracted to another method?

Comment (franz1981) [reply jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java]: This is rare, but it always happen, so I wasn't sure to pack it in a separate method: if you think it would improve readability I will do it :+1:

**Comment by franz1981** [comment]
I've spent some more time benchmarking the 3P-3C case both using `taskset` and with more forks.
This PR:
```
Benchmark                 (burstSize)  (consumerCount)  (qCapacity)                      (qType)  (warmup)  Mode  Cnt      Score     Error  Units
QueueBurstCost.burstCost          100                3       132000  MpmcUnboundedXaddArrayQueue      true  avgt  200  45400.100 ± 583.552  ns/op
```

Master:

```
Benchmark                 (burstSize)  (consumerCount)  (qCapacity)                      (qType)  (warmup)  Mode  Cnt      Score     Error  Units
QueueBurstCost.burstCost          100                3       132000  MpmcUnboundedXaddArrayQueue      true  avgt  200  46167.098 ± 592.102  ns/op
```
I can say that there is no regression although the consumer logic checks are changed.

File: jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java (hunk 3 above)
Comment (franz1981) [reply jctools-core/src/main/java/org/jctools/queues/MpmcUnboundedXaddArrayQueue.java]: Fixed ;)
